From 0afda07213ea384208256243692612b6e88a61a4 Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Thu, 11 Sep 2014 12:01:31 -0400
Subject: [ethernet] mlx4: Verify port number in __mlx4_unregister_mac

Message-id: <9f1556e72010b1a086080fe8f7686f25f5006815.1410425016.git.avadai@redhat.com>
Patchwork-id: 93689
O-Subject: [Patch RHEL7.1 V1 155/197] net/mlx4: Verify port number in __mlx4_unregister_mac
Bugzilla: 1107617 1107618 1107619 1107620
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Eugenia Emantayev <eugenia@mellanox.com>

Verify port number to avoid crashes if port number is outside the range.

Signed-off-by: Eli Cohen <eli@mellanox.com>
Signed-off-by: Eugenia Emantayev <eugenia@mellanox.com>
Signed-off-by: Amir Vadai <amirv@mellanox.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit 143b3efb402d11f64639d5729634825c3934fc4c)
Signed-off-by: Amir Vadai <avadai@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/mellanox/mlx4/port.c b/drivers/net/ethernet/mellanox/mlx4/port.c
index 5d76a60..9ba0c1c 100644
--- a/drivers/net/ethernet/mellanox/mlx4/port.c
+++ b/drivers/net/ethernet/mellanox/mlx4/port.c
@@ -244,10 +244,16 @@ EXPORT_SYMBOL_GPL(mlx4_get_base_qpn);
 
 void __mlx4_unregister_mac(struct mlx4_dev *dev, u8 port, u64 mac)
 {
- struct mlx4_port_info *info = &mlx4_priv(dev)->port[port];
- struct mlx4_mac_table *table = &info->mac_table;
+ struct mlx4_port_info *info;
+ struct mlx4_mac_table *table;
  int index;
 
+ if (port < 1 || port > dev->caps.num_ports) {
+  mlx4_warn(dev, "invalid port number (%d), aborting...\n", port);
+  return;
+ }
+ info = &mlx4_priv(dev)->port[port];
+ table = &info->mac_table;
  mutex_lock(&table->mutex);
  index = find_index(dev, table, mac);
 
-- 
1.7.1